modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java

```diff
@@ -798,51 +811,139 @@ public boolean isCheckpointApplicableForGroup(int grpId, CheckpointEntry cp) thr
-                UUID checkpointId = cp.checkpointId();
+        synchronized (earliestCpGrps) {
+            for (Integer grpId : earliestCpGrps) {
+                CacheGroupContext grpCtx = cacheGrpCtxSupplier.get(grpId);
 
-                if (data.containsKey(checkpointId))
+                if (grpCtx == null)
                     continue;
 
-                Map<Integer, GroupState> map = cp.groupStates();
+                for (GridDhtLocalPartition localPart : grpCtx.topology().localPartitions()) {
+                    long earliestCpTs = localPart.earliestCpTs();
 
-                if (map != null) {
-                    Map<Integer, GroupStateSnapshot> groupStates = new HashMap<>();
+                    if (earliestCpTs == 0)
+                        continue;
 
-                    map.forEach((k, v) ->
-                        groupStates.put(k, new GroupStateSnapshot(
-                            v.partitionIds(),
-                            v.partitionCounters(),
-                            v.size()
-                        ))
-                    );
+                    CheckpointEntry cp = histMap.get(earliestCpTs);
 
-                    data.put(checkpointId, groupStates);
+                    if (cp == null || cp.groupStates() == null || earliestCp.containsKey(cp.checkpointId()))
+                        continue;
+
+                    earliestCp.put(cp.checkpointId(), createSnapshot(cp.groupStates()));
                 }
             }
         }
 
-        Set<UUID> ids = histMap.values().stream().map(CheckpointEntry::checkpointId).collect(Collectors.toSet());
+        Set<UUID> histCpIds = histMap.values().stream()
+            .map(CheckpointEntry::checkpointId)
+            .collect(toSet());
 
-        return new EarliestCheckpointMapSnapshot(ids, data);
+        return new EarliestCheckpointMapSnapshot(histCpIds, earliestCp);
     }
 
     /**
      * Clear all cached data.
      */
     void clear() {
         histMap.clear();
-        earliestCp.clear();
+        earliestCpGrps.clear();
+    }
+
+    /** */
+    private static Map<Integer, GroupStateSnapshot> createSnapshot(Map<Integer, GroupState> stateByGrpId) {
+        Map<Integer, GroupStateSnapshot> snapshot = new HashMap<>();
+
+        for (Map.Entry<Integer, GroupState> e : stateByGrpId.entrySet()) {
+            GroupState grpState = e.getValue();
+
+            GroupStateSnapshot grpStateSnapshot = new GroupStateSnapshot(
+                grpState.partitionIds(),
+                grpState.partitionCounters(),
+                grpState.size()
+            );
+
+            snapshot.put(e.getKey(), grpStateSnapshot);
+        }
+
+        return snapshot;
+    }
+
+    /** */
+    private void clearEarliestCpTsOfGrp(int grpId) {
+        CacheGroupContext grpCtx = cacheGrpCtxSupplier.get(grpId);
+
+        if (grpCtx != null)
+            grpCtx.topology().localPartitions().forEach(localPart -> localPart.earliestCpTs(0));
```

Comment:
Don't forget about checking partition index.

Reply:
I don't quite understand what you mean, please clarify.